File: dotnet/DesignPatterns/BehavioralDesignPatterns/IteratorDesignPattern/IteratorDesignPattern.Example1/BookCollection.cs
namespace IteratorDesignPattern.Example1;

public class BookCollection : IAggregate<Book>
{
    private readonly List<Book> _books = new();

    public void AddBook(Book book)
    {
        _books.Add(book);
    }

    public int Count => _books.Count;

    public Book this[int index] => _books[index];

    public IIterator<Book> CreateIterator()
    {
        return new BookIterator(this);
    }
}

File: nodejs/DesignPatterns/BehavioralDesignPatterns/IteratorDesignPattern/IteratorDesignPattern.Example1/src/BookCollection.ts
import { Book } from "./Book";
import { IAggregate, IIterator } from "./Interfaces";
import { BookIterator } from "./BookIterator";

export class BookCollection implements IAggregate<Book> {
  private readonly _books: Book[] = [];

  addBook(book: Book): void {
    this._books.push(book);
  }

  get count(): number {
    return this._books.length;
  }

  getAt(index: number): Book {
    return this._books[index];
  }

  createIterator(): IIterator<Book> {
    return new BookIterator(this);
  }
}
